File: 07_trees/go/01_filesystem_dfs.go
package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
)

func printnames(dir string) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		log.Print(err)
		return
	}

	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Name() < entries[j].Name()
	})

	for _, entry := range entries {
		fullpath := filepath.Join(dir, entry.Name())
		fileInfo, err := os.Stat(fullpath)
		if err != nil {
			log.Print(err)
			continue
		}

		if fileInfo.Mode().IsRegular() {
			fmt.Println(entry.Name())
		} else if fileInfo.IsDir() {
			printnames(fullpath)
		}
	}
}

func main() {
	printnames("pics")
}

File: 07_trees/rust/01_filesystem_dfs.rs
use std::fs;
use std::path::Path;

fn printnames(dir: &Path) {
    let entries = match fs::read_dir(dir) {
        Ok(entries) => entries,
        Err(_) => {
            return;
        }
    };

    let mut entries: Vec<_> = entries.filter_map(|e| e.ok()).collect();
    entries.sort_by_key(|e| e.file_name());

    for entry in entries {
        let path = entry.path();
        let file_type = match entry.file_type() {
            Ok(ft) => ft,
            Err(_) => {
                continue;
            }
        };

        if file_type.is_file() {
            println!("{}", entry.file_name().to_string_lossy());
        } else if file_type.is_dir() {
            printnames(&path);
        }
    }
}

fn main() {
    printnames(Path::new("pics"));
}
